(GH-612) Add docs for upgrading to 2.0.0 (#679)
cake-contrib/Cake.Recipe: docs/input/docs/upgrading/1.x-to-2.x.md

---
Order: 10
Title: Upgrading from 1.x to 2.x
Description: Changes and suggestions needed for porting from Cake.Recipe 1.x to 2.x
---

In the 2.0.0 release of Cake.Recipe, there were a total of 15 breaking changes.  As a result, there are a number of changes that have to be made to any project that is currently using a 1.x.x release of Cake.Recipe, in order to allow it to start using this release.

Have a look at this [pull request](https://github.com/cake-contrib/Cake.DotNetVersionDetector/pull/59) on the Cake.DotNetVersionDetector repository for an example of a project that has made this transition.

## Required Changes

### Stop using username and password combination for accessing GitHub

> As of November 2020, GitHub is no longer going to support using a combination of username/password for accessing its API.  Instead, you _have_ to use a Personal Access Token.  As a result, the ability to provide a username/password via environment variables into Cake.Recipe has been removed.  Going forward, you will have to provide an environment variable for the [GITHUB_TOKEN](../fundamentals/environment-variables#github_token).

Reviewer: As in `index.cshtml`: Did you mean "\_have\_ to" or rather "_have to_"?

Author: :+1:

### Remove input parameters which are no longer required

Input parameters for the [SetParameters](../fundamentals/set-parameters) have changed.  For example the `shouldRunGitVersion` parameter is now called `shouldCalculateVersion`.

> In addition, the default values of some parameters have changed, and may no longer need to be set.  For example, `shouldRunCodecov` is now enabled by default.

Reviewer: Really not sure, but in german there'd never be a comma preceding the word "and".

Author: :+1:

> ### Change Cake.Recipe pre-processor directive to use latest version

Author: :+1:

The entry pre-processor directive should be changed to the following:

```csharp
#load nuget:?package=Cake.Recipe&version=2.0.0
```

### Pin to updated version of Cake

Cake.Recipe 1.x.x was pinned to run against version 0.32.1 of Cake.  In order to use Cake.Recipe 2.0.0 you will need to use at least 0.38.4.

### Update bootstrapper to include --bootstrap command

> Cake.Recipe now includes the use of the .Net Global Tool Module.  As such, it will be necessary to first bootstrap this module, before executing the build.  The can be done as follows:

Reviewer: ```suggestion
Cake.Recipe now includes the use of the .Net Global Tool Module.  As such, it will be necessary to first bootstrap this module, before executing the build.  That can be done as follows:
```

Author: :+1:

> Cake.Recipe now includes the use of the .Net Global Tool Module.  As such, it will be necessary to first bootstrap this module, before executing the build.  The can be done as follows:

Reviewer 2: and the comma after `module` seems wrong.
I suggest in addition to @nils-a suggestion:
```suggestion
Cake.Recipe now includes the use of the .Net Global Tool Module.  As such, it will be necessary to first bootstrap this module before executing the build.  The can be done as follows:
```

Author: :+1:

```powershell
Write-Host "Bootstrapping Cake"
dotnet cake $SCRIPT_NAME --bootstrap
if ($LASTEXITCODE -ne 0) { exit $LASTEXITCODE }

Write-Host "Running Build"
dotnet cake $SCRIPT_NAME @args
if ($LASTEXITCODE -ne 0) { exit $LASTEXITCODE }
```

### Rename AppVeyor default target

Due to the fact that Cake.Recipe is no longer directly coupled to AppVeyor, the entry target name has changed.  Instead of calling:

```yml
build_script:
  - ps: .\build.ps1 -Target AppVeyor
```

you should call:

```yml
build_script:
  - ps: .\build.ps1 --target=CI
```

### Update GitVersion.yml file with the required changes for using new version of GitVersion

Cake.Recipe has switched from using 3.6.5 of GitVersion, to 5.0.1 of the .Net Framework version, and 5.3.7 of the .Net Global Tool version.  There have been a number of breaking changes in how GitVersion configuration works, specifically around how branches are configured.  If you have done any branch configuration in the GitVersion.yml file, you may have to update to the [latest standards](https://gitversion.net/docs/configuration#branch-configuration).

> ### Ensure AppVeyor build image is using latest

Reviewer 2: ```suggestion
### Ensure AppVeyor build image is using the latest
```

Author: :+1:

In order to execute correctly on AppVeyor, you will need to use the Visual Studio 2019 build image.  Change the image definition from:

```yml
image: Visual Studio 2017
```

to:

```yml
image: Visual Studio 2019
```

### Ensure that Environment.SetVariableNames() is called

Previously, the call to [Environment.SetVariableNames](../fundamentals/set-variable-names) was not mandatory, and could be omitted from the recipe.cake file.  This is no longer the case, and it must be included.

## Suggested Changes

### Use embedded icon for NuGet package

> This guidance comes from the NuGet team directly.  Going forward, you should look to include an icon within the nupkg file.

Reviewer 2: Suggestion from Grammarly:

> Some readers see phrases like **Going forward** as stiff or just too vague. Consider rephrasing.

I tend to agree here, but not sure what would be appropriate.
The tool suggest one of the following: `In the future`, `From now on` or `As we advance`.
Neither of those feel completely right (except maybe `From now on`).

Author: :+1:

### Add CakeContrib.Guidelines Nuget Package

To help with the above suggestion, and likely others given time, we are recommending that you add the [CakeContrib.Guidelines](https://www.nuget.org/packages/CakeContrib.Guidelines/) NuGet package to your project.  This will inspect your project, and make sure that you are following the guidelines from the Cake-Contrib Organisation for your addin.

### Remove usage of nuspec file(s) for NuGet packages

> Due to the fact that Cake.Recipe has now enabled fully deterministic builds, as well as turning on of SourceLink and embedding pdb files into the generated nupkg file, it is simply easier to run `dotnet pack` rather than attempting to do this manually.  To faciliate this, you will need to add the following parameter:

Reviewer: ```suggestion
Due to the fact that Cake.Recipe has now enabled fully deterministic builds, as well as turning on of SourceLink and embedding pdb files into the generated nupkg file, it is simply easier to run `dotnet pack` rather than attempting to do this manually.  To facilitate this, you will need to add the following parameter:
```

Reviewer 2: Additionally:
```suggestion
Due to the fact that Cake.Recipe has now enabled fully deterministic builds, as well as turning on of SourceLink and embedding pdb files into the generated nupkg file, and it is simply easier to run `dotnet pack` rather than attempting to do this manually.  To faciliate this, you will need to add the following parameter:
```

Author: I have made a few additional changes to this part.

```csharp
shouldRunDotNetCorePack: true
```

### Add Cake.Addin.Analyzer package

Similar to the CakeContrib.Guidelines NuGet Package, we are also recommending that you add the [Cake.Addin.Analyzer](https://www.nuget.org/packages/Cake.Addin.Analyzer/) package to your project.  This will check through your projects and ensure that you are following best practices from a Cake perspective.

### Add settings.json file for VSCode

To help with ensuring consistency of an addins codebase, the following settings.json file is recommended:

```json
{
  "omnisharp.enableEditorConfigSupport": true,
  "omnisharp.enableRoslynAnalyzers": true,
  "[powershell]": {
    "files.encoding": "utf8bom"
  },
  "powershell.codeFormatting.addWhitespaceAroundPipe": true
}
```

### Add dependabot config file

> In order to keep your codebase up to date, we recommend that you enable dependabot to help with keeping project dependencies up to date.  This can be done by creating a `.github/dependabot.yml` file in your project with the following contents.

Reviewer 2: ```suggestion
To keep your dependencies up to date, we recommend that you enable dependabot to help with that task.  This can be done by creating a `.github/dependabot.yml` file in your project with the following contents.
```

Author: :+1:

> **NOTE:** If you choose to do this, we also recommend that you enable the `Do not build on "Pull request" events` within the AppVeyor project settings, to prevent AppVeyor from getting overwhelmed with pull requests coming from the Cake-Contrib organisation.  This is only recommended if you choose to enable GitHub Actions for your project.

Author: :+1:

### Switch to Cake.Tool rather than Cake.exe

Since Cake.Recipe now fully supports running on .Net Core, we are also recommending that you switch to use the .Net Global Tool version of Cake.  This includes using a `.config/dotnet-tools.json` file to pin to a specific version of Cake.  More information on switching to this can be found in the [Cake docs](https://cakebuild.net/docs/getting-started/setting-up-a-new-project).

### Update bootstrappers to use .Net Global Tool

An example of a build.ps1 file that can be used is:

```powershell

$ErrorActionPreference = 'Stop'

function Run([string[]]$arguments) {
    $cmd = @("& dotnet")
    $cmd += $arguments
    $cmdLine = $cmd -join " "
    Write-Verbose "> $cmdLine"
    Invoke-Expression $cmdLine

    if ($LASTEXITCODE -ne 0) {
        Write-Host "Non-Zero exit code ($($LASTEXITCODE)), exiting..."
        exit $LASTEXITCODE
	}
}

Run tool, restore

Run cake, recipe.cake, --bootstrap

$arguments = @("cake"; "recipe.cake")
$arguments += @($args)

> Run $arguments

Author: :+1:

```

An example of a build.sh file that can be used is:

```bash
#!/bin/bash
dotnet tool restore

dotnet cake recipe.cake --bootstrap

> dotnet cake recipe.cake "$@"

Author: :+1;

```

### Update appveyor.yml file for busting cache

On AppVeyor, it is possible to cache the contents of the tools folder between build executions.  Previously, this cache would have been busted using:

```yml
- tools -> recipe.cake, tools/packages.config
```

> but now that the packages.config file is not being used, it can be changed to:

Author: :+1:

```yml
- tools -> recipe.cake
```

### Delete packages.config file

> Since we are recommending switching to using the .Net Global Tool version of Cake, with associated bootstrappers, the packages.config file in the tools folder is no longer required, and can be deleted.  In addition, you should update the .gitignore file to remove the special handling of this file.

Author: Exactly, better to be safe than sorry 😄
